pyscriptjs/src/runtime.ts

```diff
@@ -0,0 +1,183 @@
+});
+
+let appConfig_: AppConfig = {
+    autoclose_loader: true,
+};
+
+appConfig.subscribe((value: AppConfig) => {
+    if (value) {
+        appConfig_ = value;
+    }
+    console.log('config set!');
+});
+
+/*
+Runtime class is a super class that all different runtimes must respect
+and adhere to.
+
+Currently, the only runtime available is Pyodide as indicated by the
+`RuntimeInterpreter` type above. This serves as a Union of types of
+different runtimes/interpreters which will be added in near future.
+
+The class has abstract methods available which each runtime is supposed
+to implement.
+
+Methods available handle loading of the interpreter, initialization,
+running code, loading and installation of packages, loading from files etc.
+
+For an example implementation, refer to the `PyodideRuntime` class
+in `pyodide.ts`
+*/
+export abstract class Runtime extends Object {
+    abstract src: string;
+    abstract name?: string;
+    abstract lang?: string;
+    abstract interpreter: RuntimeInterpreter;
+    /**
+     * global symbols table for the underlying interpreter.
+     * */
+    abstract globals: any;
+
+    /**
+     * loads the interpreter for the runtime and saves an instance of it
+     * in the `this.interpreter` property along with calling of other
+     * additional convenience functions.
+     * */
+    abstract loadInterpreter(): Promise<void>;
+
+    /**
+     * delegates the code to be run to the underlying interpreter
+     * which can call its own API behind the scenes.
+     * */
+    abstract runCode(code: string): any;
+
+    /**
+     * delegates the code to be run to the underlying interpreter
+     * (asynchronously) which can call its own API behind the scenes.
+     * */
+    abstract runCodeAsync(code: string): Promise<any>;
+
+    /**
+     * delegates the setting of JS objects to
+     * the underlying interpreter.
+     * */
+    abstract registerJsModule(name: string, module: object): void;
+
+    /**
+     * delegates the loading of packages to
+     * the underlying interpreter.
+     * */
+    abstract loadPackage(names: string | string[]): Promise<void>;
+
+    /**
+     * delegates the installation of packages
+     * (using a package manager, which can be specific to
+     * the runtime) to the underlying interpreter.
+     *
+     * For Pyodide, we use `micropip`
+     * */
+    abstract installPackage(package_name: string | string[]): Promise<void>;
+
+    /**
+     * delegates the loading of files to the
+     * underlying interpreter.
+     * */
+    abstract loadFromFile(s: string): Promise<void>;
+
+    /**
+     * initializes the page which involves loading of runtime,
+     * as well as evaluating all the code inside <py-script> tags
+     * along with initializers and postInitializers
+     * */
+    async initialize(): Promise<void> {
+        loader?.log('Loading runtime...');
+        await this.loadInterpreter();
+        const newEnv = {
+            id: 'a',
```

Comment:
What does `a` stand for..? It's not clear (and feels a bit arbitrary).

Reply:
This is also carried forward from an existing codebase, do you have any ideas here? I am thinking of `default` maybe?

Reply:
I'm not sure what it could be... although I imagine @fpliger probably knows.

I originally wondered if `id` could be something with an id defined in the DOM (e.g. `<div id="foo">...</div>`). Or it could just be some unique ID used internally by pyscipt... in which case something like a UUID or auto-incremented int might be better than a hard-coded 'a' for all case. :stuck_out_tongue:

Reply:
`a` stands for `a barf and it should be "default"` 😛 

There's a lot that we need to change on that API, the mix and match of [bad] names is one of them... In this case `id` is the ID of a specific runtime instance (that I was calling "environment"). Imho, this `id` should be `default` by default and be something else specific by the user when specifying their runtime in the app config

Reply:
changed to `default`, thanks!